amaranth/build/run.py

```diff
@@ -92,23 +87,59 @@ def execute_local(self, root="build", *, run_script=True, env=None):
                     content = content.encode("utf-8")
                 with open(filename, "wb") as f:
                     f.write(content)
-
-            if run_script:
-                if sys.platform.startswith("win32"):
-                    # Without "call", "cmd /c {}.bat" will return 0.
-                    # See https://stackoverflow.com/a/30736987 for a detailed explanation of why.
-                    # Running the script manually from a command prompt is unaffected.
-                    subprocess.check_call(["cmd", "/c", f"call {self.script}.bat"],
-                                          env=os.environ if env is None else env)
-                else:
-                    subprocess.check_call(["sh", f"{self.script}.sh"],
-                                          env=os.environ if env is None else env)
-
-            return LocalBuildProducts(os.getcwd())
-
+            return os.getcwd()
         finally:
             os.chdir(cwd)
 
+    def execute_local(self, root="build", *, run_script=True, env=None):
+        """
+        Execute build plan using the local strategy. Files from the build plan are placed in
+        the build root directory ``root``, and, if ``run_script`` is ``True``, the script
+        appropriate for the platform (``{script}.bat`` on Windows, ``{script}.sh`` elsewhere)
+        is executed in the build root. If ``env`` is not ``None``, the environment is replaced
+        with ``env``.
+
+        Returns :class:`LocalBuildProducts`.
+        """
+        build_dir = self.extract(root)
+        if run_script:
+            if sys.platform.startswith("win32"):
+                # Without "call", "cmd /c {}.bat" will return 0.
+                # See https://stackoverflow.com/a/30736987 for a detailed explanation of why.
+                # Running the script manually from a command prompt is unaffected.
+                subprocess.check_call(["cmd", "/c", f"call {self.script}.bat"],
+                                        env=os.environ if env is None else env)
+            else:
+                subprocess.check_call(["sh", f"{self.script}.sh"],
+                                        env=os.environ if env is None else env)
+
+        return LocalBuildProducts(build_dir)
+
+
+    def execute_local_docker(self, root="build", *, container_image, container_workdir="/build",
+                       docker_args=[]):
+        """
+        Execute build plan inside a docker container. Files from the build plan are placed in the 
+        build root directory ``root``, on the local filesystem. if ``run_script`` is ``True``, the
+        build root directory is is bind mounted to ``/build`` in the container and the 
+        script ``{script}.sh`` is executed inside the container.
+
+        ``docker_args`` is a dictionary that holds all arguments to docker
```

Comment:
It's not a dictionary, it's an array. Also, I would also probably just drop it. It's a single `subprocess.check_call` invocation; if you need to customize the behavior you can just do it directly. Let's not complicate lives for ourselves. I can see people wanting to e.g. remove `--rm` and that's not something this interface allows.

Reply:
for me this is necessary, since I have my vivado installation on the local disk and the most popular vivado on macos "distribution" I found does it also like this (https://github.com/ichi4096/vivado-on-silicon-mac). However, you are right, that this whole pr might be a bit superfluous now that the extract code is factored out and that might have been the proper approach all along :D.

Reply:
I am also very happy to just drop the second commit and not have any docker-specific code in amaranth